File: apps/tabtin-android/app/src/main/java/com/tabtin/mobile/features/conversation/MediaImageShapingCloud.kt
package com.tabtin.mobile.features.conversation

import kotlin.math.PI
import kotlin.math.cos
import kotlin.math.floor
import kotlin.math.hypot
import kotlin.math.max
import kotlin.math.min
import kotlin.math.roundToInt
import kotlin.math.sin

/** 逻辑空间（边长 [MediaImageShapingCloud.PRESET_SIZE]）中的一颗点。 */
internal data class ShapingDot(val x: Float, val y: Float, val r: Float)

/**
 * 「图正在成形」点云：24 个点沿闭合曲线均匀分布，曲线在 圆 → 三角 → 方 之间循环变形。
 *
 * 移植自 `packages/agent-orb/src/painter.ts` 的 `drawMorph` + `shaping` 预设（64 档）。
 * 常量、采样数、弧长重采样顺序都逐字对齐上游——差一个数就不是同一个形，
 * 一致性由 `MediaImageShapingCloudTest` 对着跨端 fixture 钉死。
 *
 * 刻意不依赖 Compose / Android framework，这样纯数学能进 JVM 单测。
 */
internal object MediaImageShapingCloud {
    /** 上游预设的逻辑边长；调用方按实际 dp 自行缩放。 */
    const val PRESET_SIZE = 64f

    /** 相位走时倍率，与 `shaping` 预设一致：调用方传入的 t 应已是「秒 × SPEED」。 */
    const val SPEED = 2.405f

    private const val ICON_D = 0.702
    private const val R_DOT = 0.008295
    private const val R_MIN = 0.25
    private const val SPREAD = 1.45
    private const val SIZE = 64.0

    private const val HOLD = 1.4
    private const val MORPH_TIME = 0.9
    private const val CYCLE = HOLD + MORPH_TIME
    private const val SAMPLES = 160

    private val SHAPES: List<(Double) -> DoubleArray> = listOf(
        { u ->
            val a = -PI / 2 + u * 2 * PI
            doubleArrayOf(cos(a) * 0.24, sin(a) * 0.24)
        },
        closedPolyline(
            arrayOf(
                doubleArrayOf(0.0, -0.26),
                doubleArrayOf(0.24, 0.16),
                doubleArrayOf(-0.24, 0.16),
            ),
        ),
        closedPolyline(
            arrayOf(
                doubleArrayOf(0.0, -0.2),
                doubleArrayOf(0.2, -0.2),
                doubleArrayOf(0.2, 0.2),
                doubleArrayOf(-0.2, 0.2),
                doubleArrayOf(-0.2, -0.2),
            ),
        ),
    )

    /**
     * 给定相位算出一帧点云。
     *
     * 全程用 Double 算、最后一步才落 Float：坐标要乘 [SIZE] 放大 64 倍，
     * Float 中间量的误差会跟着放大，容易跑出 fixture 容差。
     */
    fun dots(t: Float): List<ShapingDot> {
        val shapeCount = SHAPES.size
        val period = CYCLE * shapeCount
        // Kotlin 的 % 对负数返回负值，而上游读的是循环相位——必须归一到 [0, period)
        var c = t.toDouble() % period
        if (c < 0) c += period
        val idx = floor(c / CYCLE).toInt().coerceIn(0, shapeCount - 1)
        val local = c - idx * CYCLE
        val blend = if (local > HOLD) smoothstep((local - HOLD) / MORPH_TIME) else 0.0

        val from = SHAPES[idx]
        val to = SHAPES[(idx + 1) % shapeCount]
        val pathX = DoubleArray(SAMPLES)
        val pathY = DoubleArray(SAMPLES)
        for (i in 0 until SAMPLES) {
            val u = i.toDouble() / SAMPLES
            val a = from(u)
            val b = to(u)
            pathX[i] = (a[0] + (b[0] - a[0]) * blend) * SPREAD
            pathY[i] = (a[1] + (b[1] - a[1]) * blend) * SPREAD
        }

        val seg = DoubleArray(SAMPLES)
        var total = 0.0
        for (i in 0 until SAMPLES) {
            val j = (i + 1) % SAMPLES
            val d = hypot(pathX[j] - pathX[i], pathY[j] - pathY[i])
            seg[i] = d
            total += d
        }

        val count = max(6, (34 * ICON_D).roundToInt())
        val rDotEff = R_DOT * 1.35 * SPREAD
        val breath = 1 + 0.02 * sin(local * 3.1)
        val half = SIZE / 2
        // 两层钳制都保留：0.35 是上游绘制器的下限，R_MIN 是预设再抬一档，缺一处就跟基准对不上
        val radius = max(R_MIN, max(0.35, rDotEff * SIZE)).toFloat()

        val out = ArrayList<ShapingDot>(count)
        // cursor / walked 在整个循环里连续推进——这是沿弧长前进的游标，每点重置会退化成等参数采样
        var cursor = 0
        var walked = 0.0
        for (i in 0 until count) {
            val want = (i.toDouble() / count) * total
            while (walked + seg[cursor] < want && cursor < SAMPLES - 1) {
                walked += seg[cursor]
                cursor++
            }
            val j = (cursor + 1) % SAMPLES
            val f = if (seg[cursor] != 0.0) min(1.0, (want - walked) / seg[cursor]) else 0.0
            val px = (pathX[cursor] + (pathX[j] - pathX[cursor]) * f) * breath
            val py = (pathY[cursor] + (pathY[j] - pathY[cursor]) * f) * breath
            out.add(
                ShapingDot(
                    x = (half + px * SIZE).toFloat(),
                    y = (half + py * SIZE).toFloat(),
                    r = radius,
                ),
            )
        }
        return out
    }

    private fun smoothstep(x: Double): Double = x * x * (3 - 2 * x)

    /** 把闭合折线包成「按弧长采样」的函数，保证变形过程中点距始终均匀。 */
    private fun closedPolyline(pts: Array<DoubleArray>): (Double) -> DoubleArray {
        val n = pts.size
        val seg = DoubleArray(n)
        var total = 0.0
        for (i in 0 until n) {
            val a = pts[i]
            val b = pts[(i + 1) % n]
            val d = hypot(b[0] - a[0], b[1] - a[1])
            seg[i] = d
            total += d
        }
        return { u ->
            var want = u * total
            var i = 0
            while (want > seg[i] && i < n - 1) {
                want -= seg[i]
                i++
            }
            val a = pts[i]
            val b = pts[(i + 1) % n]
            val f = if (seg[i] != 0.0) min(1.0, want / seg[i]) else 0.0
            doubleArrayOf(a[0] + (b[0] - a[0]) * f, a[1] + (b[1] - a[1]) * f)
        }
    }
}

File: apps/tabtin-ios/Tabtin/Features/Conversation/MediaImageShapingCloud.swift
import CoreGraphics
import Foundation

/// 点云单点（逻辑坐标系，边长 `MediaImageShapingCloud.presetSize`）。
struct ShapingDot: Equatable, Sendable {
    let x: CGFloat
    let y: CGFloat
    let r: CGFloat
}

/// 「图正在成形」点云：24 个点沿闭合曲线均匀分布，曲线在 圆 → 三角 → 方 之间循环变形。
///
/// 移植自 `packages/agent-orb/src/painter.ts` 的 `drawMorph`（`shaping` 纹理 / 64 档预设）。
/// 数值与时序逐字对齐上游，跨端一致性由 `packages/agent-orb/fixtures/morph-shaping-64.json`
/// 钉死；改这里必须同时过 `MediaImageShapingCloudTests`，否则三端会画出三个「看着差不多」的形。
enum MediaImageShapingCloud {
    /// 上游逻辑画布边长。渲染层按 实际边长 / presetSize 缩放，不改这里的数。
    static let presetSize: CGFloat = 64
    /// 预设走时倍率：`dots(t:)` 的入参是「秒 × speed」后的相位。
    static let speed: CGFloat = 2.405

    // 以下四个来自 `resolveOrbPreset('shaping', 64)` 的解析结果，勿手调——
    // 它们是 BASE_OPTS 经 count / size 系数缩放后的产物，单独改一个就和上游脱钩。
    private static let iconD: CGFloat = 0.702
    private static let rDot: CGFloat = 0.008295
    private static let rMin: CGFloat = 0.25
    private static let spread: CGFloat = 1.45

    private static let hold: CGFloat = 1.4
    private static let morphTime: CGFloat = 0.9
    private static let cycle: CGFloat = hold + morphTime
    private static let shapeCount = 3
    private static let samples = 160

    /// 恒为 24。上游 `Math.round(34 * iconD)`，与渲染层无关，暴露出来供断言与预分配。
    static let dotCount: Int = max(6, Int((34 * iconD).rounded()))

    /// 减弱动效下钉死的静帧相位。
    ///
    /// 与 Electron `driveReducedMotionFrame` 的 `0.6 × preset.speed` 同值：两端必须一致，
    /// 否则同一张卡在桌面和手机上冻在不同形状，用户会当成两个不同的状态。
    static let reducedMotionPhase: CGFloat = 0.6 * speed

    private struct Point: Sendable {
        let x: CGFloat
        let y: CGFloat
    }

    /// 闭合折线的「按弧长采样」包装：变形过程中点距始终均匀，否则角点附近会堆点。
    private struct Polyline: Sendable {
        let pts: [Point]
        let seg: [CGFloat]
        let total: CGFloat

        init(_ pts: [Point]) {
            let n = pts.count
            var seg: [CGFloat] = []
            var total: CGFloat = 0
            for i in 0..<n {
                let a = pts[i]
                let b = pts[(i + 1) % n]
                let d = hypot(b.x - a.x, b.y - a.y)
                seg.append(d)
                total += d
            }
            self.pts = pts
            self.seg = seg
            self.total = total
        }

        func point(at u: CGFloat) -> Point {
            let n = pts.count
            var want = u * total
            var i = 0
            while want > seg[i] && i < n - 1 {
                want -= seg[i]
                i += 1
            }
            let a = pts[i]
            let b = pts[(i + 1) % n]
            let f = seg[i] != 0 ? min(1, want / seg[i]) : 0
            return Point(x: a.x + (b.x - a.x) * f, y: a.y + (b.y - a.y) * f)
        }
    }

    private static let triangle = Polyline([
        Point(x: 0, y: -0.26),
        Point(x: 0.24, y: 0.16),
        Point(x: -0.24, y: 0.16),
    ])

    private static let square = Polyline([
        Point(x: 0, y: -0.2),
        Point(x: 0.2, y: -0.2),
        Point(x: 0.2, y: 0.2),
        Point(x: -0.2, y: 0.2),
        Point(x: -0.2, y: -0.2),
    ])

    private static func shapePoint(index: Int, u: CGFloat) -> Point {
        switch index {
        case 0:
            let a = -CGFloat.pi / 2 + u * 2 * .pi
            return Point(x: cos(a) * 0.24, y: sin(a) * 0.24)
        case 1:
            return triangle.point(at: u)
        default:
            return square.point(at: u)
        }
    }

    private static func smoothstep(_ x: CGFloat) -> CGFloat {
        x * x * (3 - 2 * x)
    }

    /// 一帧点云。`t` 已是「秒 × `speed`」后的相位。
    static func dots(t: CGFloat) -> [ShapingDot] {
        let span = cycle * CGFloat(shapeCount)
        // Swift 的 % 对负数返回负值、对 inf/NaN 返回 NaN；两者都会让下面的 Int(floor(...)) 陷阱式崩溃。
        // 归一到 [0, span) 后负相位继续走同一轮回，非有限输入退回轮回起点。
        var c: CGFloat = 0
        if t.isFinite {
            c = t.truncatingRemainder(dividingBy: span)
            if c < 0 { c += span }
        }
        let idx = min(shapeCount - 1, max(0, Int((c / cycle).rounded(.down))))
        let local = c - CGFloat(idx) * cycle
        let blend = local > hold ? smoothstep((local - hold) / morphTime) : 0

        let next = (idx + 1) % shapeCount
        var path: [Point] = []
        path.reserveCapacity(samples)
        for i in 0..<samples {
            let u = CGFloat(i) / CGFloat(samples)
            let a = shapePoint(index: idx, u: u)
            let b = shapePoint(index: next, u: u)
            path.append(
                Point(
                    x: (a.x + (b.x - a.x) * blend) * spread,
                    y: (a.y + (b.y - a.y) * blend) * spread
                )
            )
        }

        var seg: [CGFloat] = []
        seg.reserveCapacity(samples)
        var total: CGFloat = 0
        for i in 0..<samples {
            let a = path[i]
            let b = path[(i + 1) % samples]
            let d = hypot(b.x - a.x, b.y - a.y)
            seg.append(d)
            total += d
        }

        let breath = 1 + 0.02 * sin(local * 3.1)
        let size = presetSize
        let half = size / 2
        // 两层钳制都保留：内层是 painter 的下限，外层是 collectDots 的 rMin。
        let radius = max(rMin, max(0.35, rDot * 1.35 * spread * size))

        var out: [ShapingDot] = []
        out.reserveCapacity(dotCount)
        // cursor / walked 在整个循环里连续推进——这是沿曲线走的弧长游标，每点重置会退化成等参采样。
        var cursor = 0
        var walked: CGFloat = 0
        for i in 0..<dotCount {
            let want = (CGFloat(i) / CGFloat(dotCount)) * total
            while walked + seg[cursor] < want && cursor < samples - 1 {
                walked += seg[cursor]
                cursor += 1
            }
            let a = path[cursor]
            let b = path[(cursor + 1) % samples]
            let f = seg[cursor] != 0 ? min(1, (want - walked) / seg[cursor]) : 0
            let px = (a.x + (b.x - a.x) * f) * breath
            let py = (a.y + (b.y - a.y) * f) * breath
            out.append(ShapingDot(x: half + px * size, y: half + py * size, r: radius))
        }
        return out
    }
}
